tests/samples/test_data.py

```diff
@@ -0,0 +1,208 @@
+    )
+    async def test_ok(
+        self,
+        group_setting: str,
+        data_layer: DataLayer,
+        pg: AsyncEngine,
+        fake2: DataFaker,
+        snapshot_recent,
+        spawn_client: ClientSpawner,
+        redis: Redis,
+    ):
+        client = await spawn_client(
+            authenticated=True, permissions=[Permission.create_sample]
+        )
+
+        group = await fake2.groups.create()
+
+        await data_layer.settings.update(
+            UpdateSettingsRequest(
+                sample_group=group_setting,
+                sample_all_write=True,
+                sample_group_write=True,
+            )
+        )
+        await data_layer.users.update(
+            client.user.id,
+            UpdateUserRequest(groups=[*[g.id for g in client.user.groups], group.id]),
+        )
+
+        await data_layer.users.update(
+            client.user.id,
+            UpdateUserRequest(primary_group=group.id),
+        )
+
+        jobs_client = JobsClient(redis)
+        get_data_from_app(client.app).jobs._client = jobs_client
+        get_data_from_app(client.app).samples.jobs_client = jobs_client
+
+        label = await fake2.labels.create()
+        upload = await fake2.uploads.create(user=await fake2.users.create())
+
+        await asyncio.gather(
+            client.mongo.subtraction.insert_one({"_id": "apple", "name": "Apple"}),
+        )
+
+        data = {
+            "files": [upload.id],
+            "labels": [label.id],
+            "name": "Foobar",
+            "subtractions": ["apple"],
+        }
+
+        if group_setting == "force_choice":
+            data["group"] = group.id
+
+        await client.post("/samples", data)
+
+        sample, upload = await asyncio.gather(
+            client.mongo.samples.find_one(),
+            get_row_by_id(pg, SQLUpload, 1),
+        )
+
+        assert sample == snapshot_recent(name="mongo")
+        assert await redis.lrange("jobs_create_sample", 0, -1) == [b"bf1b993c"]
+        assert upload.reserved is True
+
+
+@pytest.mark.datatest
+async def test_finalize(
+    data_layer: DataLayer,
+    spawn_job_client,
+    tmp_path,
+    get_sample_data,
+):
+    """
+    Test that sample can be finalized using the Jobs API.
+    """
+    client = await spawn_job_client(authorize=True)
+    get_config_from_app(client.app).data_path = tmp_path
+
+    await client.patch(
+        "/samples/test",
+        json={
+            "quality": {
+                "bases": [[1543]],
+                "composition": [[6372]],
+                "count": 7069,
+                "encoding": "OuBQPPuwYimrxkNpPWUx",
+                "gc": 34222440,
+                "length": [3237],
+                "sequences": [7091],
+            }
+        },
+    )
+
+    with pytest.raises(ResourceNotFoundError):
+        await get_data_from_app(client.app).uploads.get(1)
+
+    assert not (await get_row_by_id(data_layer.samples._pg, SQLSampleReads, 1)).upload
```

Comment:
* This is a mash-up of API and DL testing. You should be directly calling the data layer, not using `client`.
* Changes to the database are not verified.